lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -2861,6 +2861,63 @@ bool FoldChannelShuffle::run(Function *F, const CompilationContext &cctx) {
   return changed;
 }
 
+// Fold Tile -> Add into BatchedAdd wherever applicable.
+bool FoldTileAddIntoBatchedAdd::run(Function *F,
+                                    const CompilationContext &cctx) {
+  LOG_SCOPE(F->getLogContext(), getName());
+
+  bool changed = false;
+  for (auto &node : F->getNodes()) {
+    auto *addNode = dyn_cast<AddNode>(&node);
+    if (!addNode) {
+      continue;
+    }
+
+    NodeValue batchNode, addedNode;
+    auto &LHS = addNode->getLHS();
+    auto &RHS = addNode->getRHS();
+    TileNode *tileNode = nullptr;
+
+    // Check if LHS is a tile.
+    if ((tileNode = dyn_cast<TileNode>(LHS))) {
+      batchNode = RHS;
+      addedNode = tileNode->getInput();
+    }
+    // Check if RHS is a tile.
+    else if ((tileNode = dyn_cast<TileNode>(RHS))) {
+      batchNode = LHS;
+      addedNode = tileNode->getInput();
+      // If neither LHS or RHS is a tile, nothing to do.
+    } else {
+      continue;
+    }
+
+    // If the tiling of the added node is not along the 0th axis,
+    // Add cannot be replaced with BatchedAdd.
+    if (tileNode->getAxis() != 0) {
+      continue;
+    }
+    auto oldDims = addedNode.dims();
+    // If the 0th dimension of the added node is not 1,
+    // then reducing dimension via reshaping is more complicated.
+    // Hence, Add will not be replaced with BatchedAdd.
+    if (oldDims.size() == 0 || oldDims[0] != 1) {
+      continue;
+    }
+
+    // Reshape the added node to create a slice for the batched add
+    // such that its dim size is one less than that of the batch.
+    auto newDims = oldDims.take_back(oldDims.size() - 1);
+    auto *slice = F->createReshape("reshape", addedNode, newDims);
```

Comment:
It's a good idea to prefix/suffix this node name with whatever it's replacing so that someone can make sense of it when examining the graph. If this optimization runs multiple times, all of the `Reshape` nodes will be named `reshape`, `reshape1`, `reshape11`, ..., which is not very useful.